exporter/launch_failures.py

```diff
@@ -0,0 +1,82 @@
+from base import OSBase
+from osclient import session_adapter
+from prometheus_client import CollectorRegistry, generate_latest, Gauge
+from utils import node_details
+from datetime import datetime, timedelta
+from collections import Counter
+import logging
+
+logger = logging.getLogger(__name__)
+
+LABELS = [
+    'region',
+    'project_id'
+    'project_name']
+
+
+class LaunchFailures(OSBase):
+    """Class to report the statistics on OpenStack Nodes"""
+
+    def __init__(self, oscache, osclient):
+        super(LaunchFailures, self).__init__(oscache, osclient)
+        self.refresh_interval = oscache.refresh_interval
+
+    def build_cache_data(self):
+        """Return list of stats to cache."""
+        nova_api = session_adapter('compute')
+        keystone_api = session_adapter('identity')
+
+        nodes = node_details.get_nodes(detail=True)
+        servers = nova_api.get(
+            'servers/detail?all_tenants=True').json()['servers']
+
+        projects = keystone_api.get('v3/projects').json()['projects']
+        project_names = {p['id']: p['name'] for p in projects}
+
+        instance_failures = [
+            (s['project_id'], project_names[s['project_id']])
```

Comment:
Why not make this a `dict(project_id=s['project_id'], project_name=project_names[s['project_id']])` instead of having the `apply_labels` function? I don't think that function is adding much value and you can just create the label sets here. The amount of times the label set appears == the number of occurrences of the dict in the Counter, which should turn it into a structure like this right?

```python
{
  {'project_id': 'asdf', 'project_name': 'foo'}: 2,
  {'project_id': 'ghjk', 'project_name': 'bar'}: 1
}
```

Reply:
(Would that even work?)

Reply:
Going further, you could have a simple inline helper function:

```python
projects = keystone_api.get('v3/projects').json()['projects']
project_names = {p['id']: p['name'] for p in projects}

def labels_for(project_id):
  return dict(project_id=project_id, project_name=project_names[project_id], region=self.osclient.region)

instance_failures = [labels_for(s['project_id']) for s in servers if ... ]
node_failures = [labels_for(n.project_id) for n in nodes if ... ]
# etc.
```

Reply:
Counter can't handle unhashable types like dictionaries. Changed to a namedtuple to avoid using a labels funciton.